Finish Boost python support for DataVector

isaaclegred opened the pull request:
## Proposed changes

This Pull Request includes inserting the necessary files to add Boost functionality to SpECTRE, which will allow us to wrap SpECTRE's C++ classes in Python.    In particular it includes a wrapper class for DataVector, which will allow for the use of the DataVector class, and all of its member functions, in Python.   

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
 
-->

isaaclegred committed:
Finish Python bindings for DataVector

nilsdeppe commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -49,6 +49,46 @@ void bind_datavector() {
+      .def("asin", +[](const DataVector& t) { return DataVector{asin(t)}; })
+      .def("asinh", +[](const DataVector& t) { return DataVector{asinh(t)}; })
+      .def("atan", +[](const DataVector& t) { return DataVector{atan(t)}; })
+      .def("atan2", +[](const DataVector& t,
```
naming `t, v` `y, x` will make it easier to figure out which is which in the future. For single arg functions the name is not so important for these type of forwarding functions.

nilsdeppe commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -49,6 +49,46 @@ void bind_datavector() {
+      .def("exp10", +[](const DataVector& t) { return DataVector{exp10(t)}; })
+      .def("fabs", +[](const DataVector& t) { return DataVector{fabs(t)}; })
+      .def("hypot",
+           +[](const DataVector& t, const DataVector& v) {
```
Again here `t, v` -> `x, y`

nilsdeppe commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -49,6 +49,46 @@ void bind_datavector() {
+      .def("max", +[](const DataVector& t) { return double{max(t)}; })
+      .def("min", +[](const DataVector& t) { return double{min(t)}; })
+      .def("pow",
+           +[](const DataVector& t, double p) { return DataVector{pow(t, p)}; })
```
Here `t, p` -> `base, exp`

nilsdeppe commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -73,6 +113,7 @@ void bind_datavector() {
            +[](const DataVector& self, const double other) {
              return DataVector{other * self};
            })
+
```
no need to add this line

nilsdeppe commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -3,7 +3,7 @@
 from spectre.DataStructures import DataVector
 import unittest
-
+import math
```
Could you add back the blank line too?

nilsdeppe commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -38,5 +41,167 @@ def assignment_test():
 
 
+    def test_abs(self):
+        a=DataVector(5,-6)
```
We normally have a space between function arguments. E.g. `(5, -6)`. We also have spaces are the `=` too. 

Please apply this throughout the file.

nilsdeppe commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -38,5 +41,167 @@ def assignment_test():
+
+
+    def test_acos(self):
+        a=DataVector(5,.67)
```
Could you use `0.67` instead of `.67` so it's easier to read?

nilsdeppe commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -38,5 +41,167 @@ def assignment_test():
+
+
+    def test_tan(self):
+        a=DataVector(2,4)
```
could you make all the DataVectors be size `5` (change first argument to `5`)?

nilsdeppe reviewed (changes requested):
Let's go talk about changes, how to make a fixup commit, and the general next steps in the PR process Friday at our 1pm meeting. Overall I just have some minor stylistic comments.

nilsdeppe reviewed (commented):
I'm happy with the changes. We can go over squashing and rewording the `finding` to `binding` in the commit message next week. have a good weekend!

nilsdeppe commented:
The previous PR was merged. You'll need to cherry-pick your commit on top of that. Let's go over that together, then invite others to review next week :)

nilsdeppe reviewed (dismissed):
@nilsleiffischer if I remember correctly you asked about this PR. Would you mind giving feedback on it since you know a lot about python? Thanks!

@kidder and @wthrowe could you also review when you get some time? Thanks!

wthrowe commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -90,5 +94,257 @@ def test_output(self):
+        b = DataVector(2, 1.0)
+        b[1]= -0.5
+        self.assertEqual(b.acos()[0], 0.0)
+        self.assertEqual(b.acos()[1], math.acos(-0.5))
```
Wrong function.

nilsvu reviewed (commented):
- Add a test for `len` (raises a `TypeError` in my tests)

Additional suggestions:
- Add a binding for initialising from a list: `DataVector([1., 2., 3.])` (would also be useful for many of the tests in this PR)
- Add a binding for `__iter__`. Iterating over a `DataVector` already works because `__getitem__` is implemented, but it raises an out-of-bounds error instead of terminating nicely.
- Add a test that applies a numpy function, e.g. `np.sin(datavector)` (should work since it calls `datavector.sin()` by default)

nilsdeppe commented:
@nilsleiffischer:
- `__getitem__` raises an exception. How is this not "terminating nicely"?
- numpy apparently doesn't work as you described. `np.sin(datavector)` does out-of-bounds access and I have no idea why. Any thoughts?

nilsdeppe commented:
@nilsleiffischer we also don't know how to overload the constructor, so if you want that done you'll need to tell us how. I don't actually know if it's even possible to overload constructors in python. We should be able to add a static method that would construct the DataVector, if that's what you meant.

isaaclegred commented:
I managed to write a Static Method to create a DataVector from a Python List, is this something we would want?

nilsvu commented:
- Constructors in Python are implemented as variadic functions:
```python
def __init__(self, *args):
    if len(args) is 1:
        # construct using args[0] as an iterable
    elif len(args) is 2:
        # construct using args[0] as size and args[1] as value
    else:
        # raise error
```
Not sure how to implement this with boost, though.
- That `__getitem__` raises an out-of-bounds error is good. But what I mean is that also the following loop terminates with that exception:
```python
for x in datavector:
    print(x)
```
- `np.sin(datavector)` worked for me when I tested it. I‘ll check again.

nilsvu commented:
@isaaclegred I think it should be a constructor

isaaclegred commented:
This fixup contains a few things, first it contains the modified constructors which allow for creation from list.  It also wraps the iterator for DataVector so now the for each loop works, as well as making the DataVector functions compatible with numpy.  Also `len` is working now as well, and it has a test.

nilsdeppe commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -346,5 +351,15 @@ def test_tanh(self):
+    def test_list_constructor(self):
+        a = DataVector([1.0, 2.0, 3.0, 4.0, 5.0])
+        self.assertEqual(a[0], 1.0)
+        self.assertEqual(a[1], 2.0)
```
we should test all of them in this case.

nilsdeppe commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -346,5 +351,15 @@ def test_tanh(self):
+        a = DataVector([1.0, 2.0, 3.0, 4.0, 5.0])
+        self.assertEqual(a[0], 1.0)
+        self.assertEqual(a[1], 2.0)
 if __name__ == '__main__':
```
2 Blank lines between the class and the if

nilsdeppe commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -2,23 +2,44 @@
 // See LICENSE.txt for details.
 
 #include <boost/python.hpp>
+#include <iostream>
```
oops :) no `iostream`

nilsdeppe commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -2,23 +2,44 @@
-
 namespace bp = boost::python;
 
+// 3 "make" functions for 3 different constructors
```
Can the `make` functions be in an anonymous namespace?

isaaclegred replied:
Just to make sure this just means they are in a `namespace` with no name?

nilsdeppe replied:
Yep

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -90,5 +99,267 @@ def test_output(self):
+        a = DataVector(5, -6.0)
+        self.assertEqual(a.abs(), DataVector(5, 6.0))
+        b = DataVector(2, 1.0)
+        b[1] = -2.0
```
Use the list constructor in these tests

nilsvu commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -2,23 +2,44 @@
+      // Each "make" function gets its own constructor
+      .def("__init__", bp::make_constructor(&make_datavector_from_list))
+      .def("__init__", bp::make_constructor(&make_datavector_from_size))
+      .def("__init__", bp::make_constructor(&make_datavector_from_size_value))
```
Do you need the latter two `make_constructor`s, or can these remain `.def(bp::init<...>())` with one of them replacing the `bp::no_init`?
Also add a constructor with signature `make_datavector_from_array(bp::PyArrayObject* array)` so that we can construct from numpy arrays.

isaaclegred replied:
@nilsleiffischer  `PyArrayObject` doesn't seem to be in `boost::python`, is it possibly in a later version then we are using?

nilsvu replied:
Oh right, it‘s in numpy: https://docs.scipy.org/doc/numpy-1.13.0/reference/c-api.types-and-structures.html
We can defer this if it’s too much effort to get working.

nilsdeppe replied:
New version of Boost have Boost.numpy, but that's too new for us to be able to use, unfortunately. @osheamonn might have some insight here since he's worked with Numpy arrays in a related context before.

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -90,5 +99,267 @@ def test_output(self):
+        self.assertEqual(b.tanh()[1], math.tanh(-0.5))
+
+    #Implicitly testing the iterator
+    def test_numpy_compatibility(self):
```
How does this test the iterator? It worked in my tests before `__iter__` was added. Better add another test for the iterator.

nilsvu replied:
In this small numpy compatability test you could also also add a test along the lines of `np.array([1, 2, 3]) + DataVector([0, 1, 2]) == np.array([1, 3, 5])`

nilsdeppe replied:
Could you please give an example of what you're looking for when you say "add another test for the iterator"? Thanks!

nilsvu replied:
```python
dv = DataVector([1, 2, 3])
for i, x in enumerate(dv):
    self.assertEqual(x, dv[i])
```
This would also fail if it terminated with an out-of-bounds exception as it did before.

nilsvu reviewed (changes requested):
Cool, this is much better to work with in Python now 👍 A few refinements:

nilsvu commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -40,6 +61,48 @@ void bind_datavector() {
+      .def("log10", +[](const DataVector& t) { return DataVector{log10(t)}; })
+      .def("max", +[](const DataVector& t) { return double{max(t)}; })
+      .def("min", +[](const DataVector& t) { return double{min(t)}; })
+      .def("pow", +[](const DataVector& base,
```
This makes `dv.pow(3)` work, but you should add a `__pow__` function so that `dv**3` does the same.

nilsdeppe replied:
Is this true for all the functions? We should probably have tests that check `dv.blah()` and `blah(dv)` works.

nilsvu replied:
We only need `__pow__` to support the `**` operator. There's also a lot more special methods we could add (here's a list: https://rszalski.github.io/magicmethods/). But for math it's probably easier and sufficient to rely on numpy. We don't care about supporting `math.blah` as long as `np.blah` works (which it should already). So we could add tests for `dv.blah() == np.blah(dv)`.

Btw, a lot of non-numpy utility functions also already work because of `__iter__`, for instance `min(dv)` and `max(dv)`.

isaaclegred commented:
I believe I've addressed all the comments except for the Numpy part, which we can work out.
I didn't add explicit tests for all the constructors because they're used so often in the other tests.  The one I did test was the `size_t` constructor, because it seems that is only used one other place in the tests, though I can change this if people feel it necessary.

nilsvu commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -11,32 +10,26 @@
-      .def("__init__", bp::make_constructor(&make_datavector_from_size))
-      .def("__init__", bp::make_constructor(&make_datavector_from_size_value))
+      .def(bp::init<size_t, double>())
+      //.def("__init__", bp::make_constructor(&make_datavector_from_array))
```
delete line (we can defer adding this constructor, you can open an issue with a short description instead so that we remember this later)

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -22,6 +22,7 @@ def test_math_double(self):
         self.assertEqual(a * 2.0, DataVector(5, 2.0 * 6.7))
         self.assertEqual(a / 2.0, DataVector(5, 6.7 / 2.0))
         self.assertEqual(2.0 / a, DataVector(5, 2.0 / 6.7))
+        self.assertEqual(a ** 2.0, DataVector(5, 6.7 ** 2.0))
```
remove spaces: `a**2`

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -156,80 +142,71 @@ def test_atan2(self):
         self.assertEqual(a.exp(), DataVector(5, math.exp(1.0)))
-        b = DataVector(2, 2.0)
+        b = DataVector([2.0, -0.5])
         b[1] = -0.5
```
delete line

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -242,124 +219,119 @@ def test_hypot(self):
         self.assertEqual(a.tanh(), DataVector(5, math.tanh(4.0)))
-        b = DataVector(2, 0.0)
+        b = DataVector([0.0, -0.5])
         b[1] = -0.5
```
delete line

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -242,124 +219,119 @@ def test_hypot(self):
         self.assertEqual(np.sin(a)[1], a.sin()[1])
+        b = np.array([1.0, 2.0, 3.0])
+        c = DataVector([1.0, 2.0, 3.0])
+        self.assertEqual((b + c).all(), np.array([2.0, 4.0, 6.0]).all())
```
What is the `.all()` for? Should probably be removed

isaaclegred replied:
Python says that equality of two arrays was ambiguous, and it must be explicitly stated whether if it is equality for any element or equality for all elements.  The test fails if it is removed.

nilsvu replied:
Either use `assertSequenceEqual` (preferred imo) or assert for `(b + c == np.array(...)).all()`. What you currently do reduces `(b + c)` to a boolean, which is true only if all elements are 1. So the test succeeds, but it doesn’t test what we want.

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -242,124 +219,119 @@ def test_hypot(self):
+
     def test_numpy_compatibility(self):
         a = DataVector(5, 2.0)
         self.assertEqual(np.sin(a)[1], a.sin()[1])
```
You could do:
```python
x = np.linspace(0, 2 * np.pi, 10)
self.assertEqual(DataVector(list(x)).sin(), np.sin(x))
```

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -242,124 +219,119 @@ def test_hypot(self):
         a = DataVector([1.0, 2.0, 3.0, 4.0, 5.0])
-        self.assertEqual(a[0], 1.0)
-        self.assertEqual(a[1], 2.0)
+        for i,val in enumerate(a):
```
add space: `i, val`

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -242,124 +219,119 @@ def test_hypot(self):
-        self.assertEqual(a[0], 1.0)
-        self.assertEqual(a[1], 2.0)
+        for i,val in enumerate(a):
+            self.assertEquals(val, a[i])
```
does `assertEquals` exist or is it a typo (have you run these tests)?

isaaclegred replied:
It was a typo sorry, but I have run the tests so I guess it exists, I'll change it though.

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -242,124 +219,119 @@ def test_hypot(self):
+        self.assertEqual((b + c).all(), np.array([2.0, 4.0, 6.0]).all())
 
 
-    def test_list_constructor(self):
```
I think you should actually test the list constructor by itself. The reason for having these fine-grained unit tests is so that we are able to narrow down precisely what goes wrong if something breaks.

isaaclegred commented:
These commits I believe address all of the problems above, the fixup is basically just fixing spacing in the tests.   I couldn't figure out how to get `assertSequenceEqual` to work, so I used the second way you suggested, @nilsleiffischer, with the `assert` to test the equality of the arrays.  Also I re-added `test_list_constructor`.  Any other comments?  Anything else you want to see fixed @nilsdeppe ?

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -16,6 +22,7 @@ def test_math_double(self):
+        self.assertEqual(a**2.0, DataVector(5, 6.7**2.0))
 
         b = DataVector(2)
         b[0] = 7.2
```
change this to use the list constructor too, while you're at it.

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -81,14 +85,290 @@ def assignment_test():
+        self.assertEqual(np.sin(a)[1], a.sin()[1])
+        b = np.array([1.0, 2.0, 3.0])
+        c = DataVector([1.0, 2.0, 3.0])
+        assert((b + c) == np.array([2.0, 4.0, 6.0])).all()
```
`self.assertTrue`

nilsvu commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -81,14 +85,290 @@ def assignment_test():
+        c = DataVector([1.0, 2.0, 3.0])
+        assert((b + c) == np.array([2.0, 4.0, 6.0])).all()
+        x = np.linspace(0, 2 * np.pi, 10)
+        assert(DataVector(list(x)).sin() == np.sin(x)).all()
```
`self.assertTrue` and delete the duplicate `sin` test above

nilsdeppe commented:
@isaaclegred you have a few errors to fix before we should look at this again :)

nilsvu reviewed (dismissed):
Looks good to me 👍 Are the Travis segfaults known @nilsdeppe?

nilsdeppe commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -40,10 +53,56 @@ void bind_datavector() {
            +[](const DataVector& self, const double other) {
              return DataVector{self + other};
            })
+      .def("__pow__",
```
It would make more sense to put `__pow__` before `__add__`

nilsdeppe commented on tests/Unit/DataStructures/Test_DataVector.py:
```diff
@@ -81,14 +83,288 @@ def assignment_test():
+    def test_hypot(self):
+        a = DataVector(5, 3.0)
+        b = DataVector(5, 4.0)
+        self.assertEqual(a.hypot(b)[0], 5.0)
```
Is there a reason not to have a different element test for `hypot` like you have for all the other functions?

isaaclegred replied:
No, sorry, when I wrote all the other ones I was going to ask whether you wanted the tests for this and `atan2` to be two `DataVector`'s each with different elements or if you wanted only one to have differing elements but then I forgot.  I'll just add in the most general cases.

nilsdeppe reviewed (changes requested):
A couple minor things that I'm happy for you to squash right away.

@nilsleiffischer yes, those are the "usual" segfaults (I restarted the build)

@wthrowe @kidder could you please look at this when you get a chance? Thanks :)
